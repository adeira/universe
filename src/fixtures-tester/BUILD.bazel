load("//:bazel/macros/build_yarn_workspace.bzl", "build_yarn_workspace")
load("//:bazel/macros/test_yarn_workspace.bzl", "test_yarn_workspace")

build_yarn_workspace(
    name = "fixtures-tester",
    package_name = "@adeira/fixtures-tester",
)

test_yarn_workspace(
    name = "test",
)
