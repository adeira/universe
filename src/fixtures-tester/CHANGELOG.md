# Unreleased

# 1.1.0

- New function `verifyTestsFromFixtures` added - it has similar interface as `generateTestsFromFixtures` except it verifies that the operation result is `true` instead of creating Jest snapshots. This is useful when you want to run some tests on your fixtures files but there is not much to snapshot.

# 1.0.1

Note for all Flow users: all projects in [`adeira/universe`](https://github.com/adeira/universe) now use implicit exact Flow types (`{}` for strict objects and `{ ... }` for open objects, syntax `{||}` is deprecated). We do not expect any issues as long as you are using `exact_by_default=true` Flow option.

- Fix failure on unexpected subfolders, see: https://github.com/adeira/universe/pull/2257

# 1.0.0

Initial release.
