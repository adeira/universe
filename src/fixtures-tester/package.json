{
  "name": "@adeira/fixtures-tester",
  "description": "Utility to ease testing with fixtures.",
  "homepage": "https://github.com/adeira/fixtures-tester",
  "bugs": "https://github.com/adeira/universe/issues",
  "repository": {
    "type": "git",
    "url": "git@github.com:adeira/universe.git",
    "directory": "src/fixtures-tester"
  },
  "license": "MIT",
  "private": false,
  "version": "1.1.1",
  "main": "./src/index.js",
  "type": "commonjs",
  "sideEffects": false,
  "dependencies": {
    "@adeira/js": "^2.1.1",
    "@babel/runtime": "^7.16.7"
  }
}
