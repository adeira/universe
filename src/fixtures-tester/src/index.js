// @flow

export { default as generateTestsFromFixtures } from './generateTestsFromFixtures';
export { default as verifyTestsFromFixtures } from './verifyTestsFromFixtures';
