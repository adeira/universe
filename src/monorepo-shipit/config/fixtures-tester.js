// @flow strict

import type { ConfigType } from '../ConfigType.flow';

module.exports = ({
  getStaticConfig() {
    return {
      repository: 'git@github.com:adeira/fixtures-tester.git',
    };
  },
  getPathMappings() {
    return new Map([['src/fixtures-tester/', '']]);
  },
}: ConfigType);
