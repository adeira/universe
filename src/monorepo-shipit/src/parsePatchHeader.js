// @flow strict

import Changeset from './Changeset';
import splitHead from './splitHead';

export default function parsePatchHeader(header: string): Changeset {
  const [rawEnvelope, rawBody] = splitHead(header, '\n\n');

  const description = [];
  const coAuthorLines = [];

  // Co-authored-by must be the absolute last thing in the message so we separate it here from the
  // description and compose it later correctly (to add the `adeira-source-id` label correctly).
  for (const line of rawBody.trim().split('\n')) {
    if (line.startsWith('Co-authored-by:')) {
      coAuthorLines.push(line);
    } else {
      description.push(line);
    }
  }

  let changeset = new Changeset()
    .withDescription(description.join('\n'))
    .withCoAuthorLines(coAuthorLines);

  const envelope = rawEnvelope.replace(/(?:\n\t|\n )/, ' ');
  for (const line of envelope.split('\n')) {
    if (!line.includes(':')) {
      continue;
    }
    const [key, rawValue] = splitHead(line, ':');
    const value = rawValue.trim();
    switch (key.trim().toLowerCase()) {
      case 'from':
        changeset = changeset.withAuthor(value);
        break;
      case 'subject':
        changeset = changeset.withSubject(value.replace(/^\[PATCH] /, ''));
        break;
      case 'date':
        changeset = changeset.withTimestamp(value);
        break;
    }
  }
  return changeset;
}
