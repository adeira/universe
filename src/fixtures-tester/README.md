Install (via Yarn or NPM):

```text
yarn add @adeira/fixtures-tester
```

## `generateTestsFromFixtures`

This function loads all the fixture files (their file content) from your folder and snapshots the result of your callback operation. This is handy when you have several example files, you want to perform some operation on top of them and snapshot the results.

```js
import { generateTestsFromFixtures } from '@adeira/fixtures-tester';

generateTestsFromFixtures(
  `${__dirname}/__fixtures__`,
  (input: string): string => {
    return doSomethingWithTheInput(input); // this result will be saved in a Jest snapshot
  },
  // optionally a snapshot name
);
```

There are two additional modes which allow you to control the fixtures behavior more precisely. There are based on the fixture name:

- `simpleExample.graphql` - fixture will snapshot as expected
- `simpleExample.only.graphql` - only these snapshots will run (for debug only)
- `simpleExample.error.graphql` - it's expected that fixture will throw an error (expected to fail tests)

The error file convention is required to prevent unintentional snapshots of exceptions.

## `verifyTestsFromFixtures`

This function simply verifies that the callback returns `true` but it doesn't snapshot anything:

```js
import { verifyTestsFromFixtures } from '@adeira/fixtures-tester';

verifyTestsFromFixtures(`${__dirname}/__fixtures__`, (input: string): boolean => {
  return isInputValid(input) === true;
});
```

This is handy when the snapshot from `generateTestsFromFixtures` would be meaningless but you still want to check all available fixtures.
