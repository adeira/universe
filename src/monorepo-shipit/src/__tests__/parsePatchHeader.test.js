// @flow strict-local

import path from 'path';
import { generateTestsFromFixtures } from '@adeira/fixtures-tester';

import parsePatchHeader from '../parsePatchHeader';

generateTestsFromFixtures(path.join(__dirname, 'fixtures', 'headers'), parsePatchHeader);
