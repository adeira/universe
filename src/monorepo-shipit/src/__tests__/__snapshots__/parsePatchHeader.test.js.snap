// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: co-authored-by.header 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
From 160feaf6d2637b22216987d1f95b96d585a838a6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Zl=C3=A1mal?= <mrtnzlml@gmail.com>
Date: Thu, 15 Apr 2021 13:18:49 -0500
Subject: [PATCH] SX Design: add Norwegian language

Co-authored-by: Trond Bergquist <trond_bergquist@hotmail.com>


~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "coAuthorLines": [
    "Co-authored-by: Trond Bergquist <trond_bergquist@hotmail.com>"
  ],
  "debugMessages": [],
  "description": "",
  "author": "=?UTF-8?q?Martin=20Zl=C3=A1mal?= <mrtnzlml@gmail.com>",
  "timestamp": "Thu, 15 Apr 2021 13:18:49 -0500",
  "subject": "SX Design: add Norwegian language"
}
`;

exports[`matches expected output: co-authored-by-multiple.header 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
From 160feaf6d2637b22216987d1f95b96d585a838a6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Zl=C3=A1mal?= <mrtnzlml@gmail.com>
Date: Thu, 15 Apr 2021 13:18:49 -0500
Subject: [PATCH] SX Design: add Norwegian language

Co-authored-by: Trond Bergquist <trond_bergquist@hotmail.com>
Co-authored-by: Patricia Bergquist <patricia_bergquist@hotmail.com>


~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "coAuthorLines": [
    "Co-authored-by: Trond Bergquist <trond_bergquist@hotmail.com>",
    "Co-authored-by: Patricia Bergquist <patricia_bergquist@hotmail.com>"
  ],
  "debugMessages": [],
  "description": "",
  "author": "=?UTF-8?q?Martin=20Zl=C3=A1mal?= <mrtnzlml@gmail.com>",
  "timestamp": "Thu, 15 Apr 2021 13:18:49 -0500",
  "subject": "SX Design: add Norwegian language"
}
`;

exports[`matches expected output: multiline-subject.header 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
From 93ee65de02e77c52eb9d677b859554b40253bafa Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Zl=C3=A1mal?= <adeira@example.com>
Date: Wed, 10 Apr 2019 16:50:47 -0500
Subject: [PATCH] Docs: add Relay and GraphQL related videos from our internal
 JS Saturday


~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "coAuthorLines": [],
  "debugMessages": [],
  "description": "",
  "author": "=?UTF-8?q?Martin=20Zl=C3=A1mal?= <adeira@example.com>",
  "timestamp": "Wed, 10 Apr 2019 16:50:47 -0500",
  "subject": "Docs: add Relay and GraphQL related videos from our internal JS Saturday"
}
`;

exports[`matches expected output: simple.header 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
From bf0a8783cf416f796659e60fe735bba98e676e67 Mon Sep 17 00:00:00 2001
From: Automator <adeira@example.com>
Date: Mon, 4 Feb 2019 13:29:04 -0600
Subject: [PATCH] Import and publish '@adeira/relay' (0.1.0)

Formerly known as '@mrtnzlml/relay'


~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "coAuthorLines": [],
  "debugMessages": [],
  "description": "Formerly known as '@mrtnzlml/relay'",
  "author": "Automator <adeira@example.com>",
  "timestamp": "Mon, 4 Feb 2019 13:29:04 -0600",
  "subject": "Import and publish '@adeira/relay' (0.1.0)"
}
`;

exports[`matches expected output: unicode.header 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20Zl=C3=A1mal?= <adeira@example.com>
Date: Wed, 10 Apr 2019 16:50:47 -0500
Subject: [PATCH] Let's party 🍺 🍻 🍷 🍸😀 😬 😁 😂 😃 😄 😅


~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "coAuthorLines": [],
  "debugMessages": [],
  "description": "",
  "author": "=?UTF-8?q?Martin=20Zl=C3=A1mal?= <adeira@example.com>",
  "timestamp": "Wed, 10 Apr 2019 16:50:47 -0500",
  "subject": "Let's party 🍺 🍻 🍷 🍸😀 😬 😁 😂 😃 😄 😅"
}
`;
